File: rbql-py/rbql/rbql_json.py
import json
import sys
import os
import io
import re

from . import rbql_engine
from . import csv_utils
from . import rbql_csv

debug_mode = False

def set_debug_mode():
    global debug_mode
    debug_mode = True


def get_json_object_to_write(header, fields):
    if len(fields) == 1:
        return fields[0]
    result = dict()
    for i in range(len(fields)):
        key_name = header[i] if i < len(header) else 'col{}'.format(i)
        result[key_name] = fields[i]
    return result


def finalize_stream(stream, close_stream_on_finish):
    if close_stream_on_finish:
        stream.close()
        return
    try:
        stream.flush()
    except BrokenPipeError as exc:
        try:
            sys.stdout.close()
        except Exception:
            pass


def deduplicate_header_keys(header):
    # This algorithm is O(N^2) but we don't expect to have a lot of keys.
    # FIXME unit-test this function
    unique_keys = list()
    deduplicated_keys = set()
    for h in header:
        unique_key = h
        dedup_counter = 1
        while unique_key in unique_keys:
            dedup_counter += 1
            unique_key = '{}_{}'.format(h, dedup_counter)
        assert unique_key not in unique_keys
        unique_keys.append(unique_key)
        if dedup_counter != 1:
            deduplicated_keys.add(h)
    return (unique_keys, deduplicated_keys)


class JsonLinesWriter(rbql_engine.RBQLOutputWriter):
    def __init__(self, stream, close_stream_on_finish, encoding, line_separator='\n'):
        assert encoding in ['utf-8', 'latin-1', None]
        self.stream = rbql_csv.encode_output_stream(stream, encoding)
        self.line_separator = line_separator
        self.close_stream_on_finish = close_stream_on_finish
        self.broken_pipe = False
        self.header = []

    def write(self, fields):
        object_to_write = get_json_object_to_write(self.header, fields)
        try:
            json_str = json.dumps(object_to_write, ensure_ascii=False, default=str)
        except TypeError as e:
            raise rbql_engine.RbqlIOHandlingError('Error serializing object to JSON: {}'.format(e))

        try:
            self.stream.write(json_str)
            self.stream.write(self.line_separator)
            return True
        except BrokenPipeError as exc:
            self.broken_pipe = True
            return False

    def finish(self):
        if self.broken_pipe:
            return
        finalize_stream(self.stream, self.close_stream_on_finish)

    # FIXME apparently this not always gets called, we should mark all json files to have headers like csv `with headers` flag.
    # FIXME make sure it works with multistep paths
    def set_header(self, header):
        if header is None:
            return
        # Json objects don't allow duplicate keys so we have to dedup them to make sure they are unique to avoid accidental data loss.
        self.header, self.deduplicated_keys = deduplicate_header_keys(header)

    def get_warnings(self):
        warnings = []
        if len(self.deduplicated_keys) != 0:
            sorted_keys = sorted(['"{}"'.format(v) for v in list(self.deduplicated_keys)])
            warnings.append('Deduplicated output json keys to avoid data loss: {}'.format(', '.join(sorted_keys)))
        return warnings


# FIXME add a unit test with 2 columns with the same name - json should skip them.
# FIXME since json skips columns with the same name consider implementing column deduplication in the output header to avoid silently dropping them.


class JsonArrayObjectRecordIterator(rbql_engine.RBQLInputIterator):
    def __init__(self, stream, encoding, table_name='input', variable_prefix='a'):
        assert encoding in ['utf-8', 'latin-1', None]
        self.encoding = encoding
        self.stream = rbql_csv.encode_input_stream(stream, encoding)
        self.table_name = table_name
        self.variable_prefix = variable_prefix

        self.NR = 0 # Record number
        self.NL = 0 # Line number
        try:
            self.json_object = json.load(self.stream)
        except json.decoder.JSONDecodeError as e:
            raise rbql_engine.RbqlIOHandlingError('Unable to parse input as JSON: {}'.format(e))
        if not isinstance(self.json_object, list):
            raise rbql_engine.RbqlIOHandlingError('Input JSON root node must be array in array iteration mode')
        self.NL = 1 # The object has to start at the first line so we just keep NL at 1.

    def get_header(self):
        # FIXME consider if this is a hack or not. Test queries with stars like `SELECT a.*, a.*` or `SELECT *`.
        # We might not need this (i.e. we can have it return None as the default impl) if we support the write_header flag, see the rbql_engine.py comments.
        return [self.variable_prefix + '1']

    def get_record(self):
        if self.NR >= len(self.json_object):
            return None
        self.NR += 1
        return [self.json_object[self.NR - 1]]


class JsonArrayObjectWriter(rbql_engine.RBQLOutputWriter):
    def __init__(self, stream, close_stream_on_finish, encoding, line_separator='\n', pretty_indent=None):
        assert encoding in ['utf-8', 'latin-1', None]
        self.stream = rbql_csv.encode_output_stream(stream, encoding)
        self.line_separator = line_separator
        self.close_stream_on_finish = close_stream_on_finish
        self.broken_pipe = False
        self.header = []
        self.pretty_indent=pretty_indent
        self.num_records_written = 0
        self.deduplicated_keys = set()

    def write(self, fields):
        object_to_write = get_json_object_to_write(self.header, fields)
        try:
            # Intentionally do not split json_str to add extra pretty indents because the root level is a flat array and shifting everything right just reduces density for the sake of dubious consistency. 
            json_str = json.dumps(object_to_write, ensure_ascii=False, default=str, indent=self.pretty_indent)
        except TypeError as e:
            raise rbql_engine.RbqlIOHandlingError('Error serializing object to JSON: {}'.format(e))

        try:
            if self.num_records_written == 0:
                self.stream.write('[')
                self.stream.write(self.line_separator)
            else:
                self.stream.write(',')
                self.stream.write(self.line_separator)
            self.stream.write(json_str)
            self.num_records_written += 1
            return True
        except BrokenPipeError as exc:
            self.broken_pipe = True
            return False

    def finish(self):
        if self.broken_pipe:
            return
        if self.num_records_written == 0:
            # Output an empty array if no entries were produced
            self.stream.write('[')
        self.stream.write(self.line_separator)
        self.stream.write(']')
        self.stream.write(self.line_separator) # POSIX requires a newline at the end of text files.
        finalize_stream(self.stream, self.close_stream_on_finish)

    # FIXME apparently this not always gets called, we should mark all json files to have headers like csv `with headers` flag.
    # FIXME make sure it works with multistep paths
    def set_header(self, header):
        if header is None:
            return
        # Json objects don't allow duplicate keys so we have to dedup them to make sure they are unique to avoid accidental data loss.
        self.header, self.deduplicated_keys = deduplicate_header_keys(header)

    def get_warnings(self):
        warnings = []
        if len(self.deduplicated_keys) != 0:
            sorted_keys = sorted(['"{}"'.format(v) for v in list(self.deduplicated_keys)])
            warnings.append('Deduplicated output json keys to avoid data loss: {}'.format(', '.join(sorted_keys)))
        return warnings


# NOTE: using json lines format as input is essentially equivalent to `select json.loads(a1) | select a1['name']` type of query.
class JsonLinesRecordIterator(rbql_engine.RBQLInputIterator):
    def __init__(self, stream, encoding, table_name='input', variable_prefix='a', chunk_size=1024):
        assert encoding in ['utf-8', 'latin-1', None]
        self.encoding = encoding
        self.stream = rbql_csv.encode_input_stream(stream, encoding)
        self.table_name = table_name
        self.variable_prefix = variable_prefix

        self.buffer = ''
        self.exhausted = False
        self.NR = 0 # Record number
        self.NL = 0 # Line number
        self.chunk_size = chunk_size
        self.utf8_bom_removed = False

    def get_header(self):
        # FIXME consider if this is a hack or not. Test queries with stars like `SELECT a.*, a.*` or `SELECT *`.
        # We might not need this (i.e. we can have it return None as the default impl) if we support the write_header flag, see the rbql_engine.py comments.
        return [self.variable_prefix + '1']

    def _get_row_from_buffer(self):
        str_before, separator, str_after = csv_utils.extract_line_from_data(self.buffer)
        if separator is None:
            return None
        if separator == '\r' and str_after == '':
            one_more = self.stream.read(1)
            if one_more == '\n':
                separator = '\r\n'
            else:
                str_after = one_more
        self.buffer = str_after
        return str_before

    def _read_until_found(self):
        if self.exhausted:
            return
        chunks = []
        while True:
            chunk = self.stream.read(self.chunk_size)
            if not chunk:
                self.exhausted = True
                break
            chunks.append(chunk)
            if csv_utils.newline_rgx.search(chunk) is not None:
                break
        self.buffer += ''.join(chunks)

    def get_row(self):
        try:
            row = self._get_row_from_buffer()
            if row is None:
                self._read_until_found()
                row = self._get_row_from_buffer()
                if row is None:
                    assert self.exhausted
                    if not len(self.buffer):
                        return None
                    row = self.buffer
                    self.buffer = ''
            self.NL += 1
            if self.NL == 1:
                clean_line = rbql_csv.remove_utf8_bom(row, self.encoding)
                if clean_line != row:
                    row = clean_line
                    self.utf8_bom_removed = True
            return row
        except UnicodeDecodeError:
            raise rbql_engine.RbqlIOHandlingError('Unable to decode input table as UTF-8. Use binary (latin-1) encoding instead')

    def get_record(self):
        while True:
            line = self.get_row()
            if line is None:
                return None
            line = line.strip()
            if not line:
                continue
            try:
                json_obj = json.loads(line)
                self.NR += 1
                return [json_obj]
            except json.JSONDecodeError as e:
                raise rbql_engine.RbqlIOHandlingError('Error decoding JSON in {} table at record {}, line {}: {}'.format(self.table_name, self.NR + 1, self.NL, str(e)))

    def get_warnings(self):
        result = []
        if self.utf8_bom_removed:
            result.append('UTF-8 Byte Order Mark (BOM) was found and skipped in {} table'.format(self.table_name))
        return result


# TODO we might want the output to optionally be CSV too. 
def query_json(query_text, input_path, output_path, output_warnings, user_init_code='', input_json_lines=True, output_json_lines=True, pretty_indent=None):
    output_stream, close_output_on_finish = (None, False)
    input_stream, close_input_on_finish = (None, False)
    join_tables_registry = None
    try:
        output_stream, close_output_on_finish = (sys.stdout, False) if output_path is None else (open(output_path, 'wb'), True)
        input_stream, close_input_on_finish = (sys.stdin, False) if input_path is None else (open(input_path, 'rb'), True)

        default_init_source_path = os.path.join(os.path.expanduser('~'), '.rbql_init_source.py')
        if user_init_code == '' and os.path.exists(default_init_source_path):
            user_init_code = rbql_csv.read_user_init_code(default_init_source_path)
        input_iterator = None
        if input_json_lines:
            input_iterator = JsonLinesRecordIterator(input_stream, 'utf-8', table_name='input', variable_prefix='a')
        else:
            input_iterator = JsonArrayObjectRecordIterator(input_stream, 'utf-8', table_name='input', variable_prefix='a')
        if output_json_lines:
            output_writer = JsonLinesWriter(output_stream, close_output_on_finish, 'utf-8')
        else:
            output_writer = JsonArrayObjectWriter(output_stream, close_output_on_finish, 'utf-8', pretty_indent=pretty_indent)
        if debug_mode:
            rbql_engine.set_debug_mode()
        rbql_engine.query(query_text, input_iterator, output_writer, output_warnings, join_tables_registry, user_init_code)
    finally:
        if close_input_on_finish:
            input_stream.close()
        if close_output_on_finish:
            output_stream.close()

File: rbql-js/rbql_json.js
const fs = require('fs');
const os = require('os');
const path = require('path');
const util = require('util');

const rbql = require('./rbql.js');
const csv_utils = require('./csv_utils.js');

class RbqlIOHandlingError extends Error {}


// FIXME add json file test with unicode chars. For py and js versions. Use both input stream and input path modes.

function assert(condition, message=null) {
    if (!condition) {
        if (!message) {
            message = 'Assertion error';
        }
        throw new AssertionError(message);
    }
}


function deduplicate_header_keys(header) {
    // This algorithm is O(N^2) but we don't expect to have a lot of keys.
    // FIXME add unit tests for this.
    let unique_keys = [];
    let deduplicated_keys = new Set();
    for (let h of header) {
        let unique_key = h;
        let dedup_counter = 1;
        while (unique_keys.indexOf(unique_key) != -1) {
            dedup_counter += 1;
            unique_key = `${h}_${dedup_counter}`;
        }
        unique_keys.push(unique_key);
        if (dedup_counter != 1)
            deduplicated_keys.add(h);
    }
    return [unique_keys, Array.from(deduplicated_keys).sort()];
}


function get_json_object_to_write(header, fields) {
    // FIXME add unit tests.
    if (fields.length == 1)
        return fields[0];
    if (header.length && fields.length != header.length)
        throw new RbqlIOHandlingError(`Inconsistent number of columns in output header and the current record: ${header.length} != ${fields.length}`);
    let result = {};
    for (let i = 0; i < fields.length; i++) {
        let key_name = i < header.length ? header[i] : `col_${i + 1}`;
        result[key_name] = fields[i];
    }
    return result;
}


class JsonLinesWriter extends rbql.RBQLOutputWriter {
    constructor(stream, close_stream_on_finish, encoding='utf-8', line_separator='\n') {
        super();
        this.stream = stream;
        this.encoding = encoding;
        if (encoding)
            this.stream.setDefaultEncoding(encoding);
        this.stream.on('error', (error_obj) => { this.store_first_error(error_obj); })
        this.line_separator = line_separator;

        this.close_stream_on_finish = close_stream_on_finish;

        this.header = [];
        this.deduplicated_keys = [];
        this.first_error = null;
    }

    store_first_error(error_obj) {
        // Store only first error because it is typically more important than the subsequent ones.
        if (this.first_error === null)
            this.first_error = error_obj;
    }

    set_header(header) {
        if (header !== null) {
            [this.header, this.deduplicated_keys] = deduplicate_header_keys(header);
        }
    }

    async finish() {
        let close_stream_on_finish = this.close_stream_on_finish;
        let output_stream = this.stream;
        let output_encoding = this.encoding;
        let writer_error = this.first_error;
        let finish_promise = new Promise(function(resolve, reject) {
            if (writer_error !== null) {
                reject(writer_error);
            }
            if (close_stream_on_finish) {
                output_stream.end('', output_encoding, () => { resolve(); });
            } else {
                setTimeout(() => { resolve(); }, 0);
            }
        });
        return finish_promise;
    };


    get_warnings() {
        let result = [];
        if (this.deduplicated_keys.length > 0) {
            let keys_to_report = this.deduplicated_keys.map(key => `"${key}"`).join(', ');
            result.push(`Deduplicated output json keys to avoid data loss: ${keys_to_report}`);
        }
        return result;
    };


    async do_write(object_to_write) {
        this.stream.write(JSON.stringify(object_to_write));
        this.stream.write(this.line_separator);
        let writer_error = this.first_error;
        return new Promise(function(resolve, reject) {
            if (writer_error !== null) {
                reject(writer_error);
            } else {
                resolve();
            }
        });
    }


    async write(fields) {
        let object_to_write = get_json_object_to_write(this.header, fields);
        await this.do_write(object_to_write);
        // The return value is needed for the stacked writers architecture in rbql.js.
        // So far the only writer that can return `false` is the TopWriter which uses it as a flag to request stop when the limit is reached.
        // This writer always succeeds (unless there is an exception).
        return true;
    };
}


class JsonLinesRecordIterator extends rbql.RBQLInputIterator {
    // TODO add query modifier with "noheaders" this would name keys as `a1`, `a2`, etc.
    // FIXME adjust
    // FIXME add unit tests
    constructor(stream, encoding='utf-8', table_name='input', variable_prefix='a') {
        super();
        this.stream = stream;
        this.encoding = encoding;
        this.table_name = table_name;
        this.variable_prefix = variable_prefix;


        this.decoder = null;
        if (encoding == 'utf-8') {
            // This was copied from the csv impl, see comments there.
            this.decoder = new util.TextDecoder(encoding, {fatal: true, stream: true});
        }

        this.input_exhausted = false;
        this.started = false;

        this.NR = 0; // Record number
        this.NL = 0; // Line number

        this.partially_decoded_line = '';
        this.partially_decoded_line_ends_with_cr = false;

        // Holds an external "resolve" function which is called when everything is fine.
        this.resolve_current_record = null;
        // Holds an external "reject" function which is called when error has occured.
        this.reject_current_record = null;
        // Holds last exception if we don't have any reject callbacks from clients yet.
        this.current_exception = null;

        this.produced_records_queue = new csv_utils.RecordQueue();
    }

    get_header() {
        // FIXME consider if this is a hack or not. Test queries with stars like `SELECT a.*, a.*` or `SELECT *`.
        // We might not need this (i.e. we can have it return None as the default impl) if we support the write_header flag, see the rbql_engine.py comments.
        return [`${this.variable_prefix}1`];
    }

    reset_external_callbacks() {
        // Drop external callbacks simultaneously since promises can only resolve once, see: https://stackoverflow.com/a/18218542/2898283
        this.reject_current_record = null;
        this.resolve_current_record = null;
    }

    try_propagate_exception() {
        if (this.current_exception && this.reject_current_record) {
            let reject = this.reject_current_record;
            let exception = this.current_exception;
            this.reset_external_callbacks();
            this.current_exception = null;
            reject(exception);
        }
    }


    store_or_propagate_exception(exception) {
        if (this.current_exception === null)
            // Ignore subsequent exceptions if we already have an unreported error. This way we prioritize earlier errors over the more recent ones.
            this.current_exception = exception;
        this.try_propagate_exception();
    }


    try_resolve_next_record() {
        this.try_propagate_exception();
        if (this.resolve_current_record === null)
            return;

        let record = this.produced_records_queue.dequeue();
        if (record === null && !this.input_exhausted)
            return;
        let resolve = this.resolve_current_record;
        this.reset_external_callbacks();
        resolve(record);
    };


    async get_record() {
        if (!this.started)
            await this.start();
        if (this.stream && this.stream.isPaused())
            this.stream.resume();

        let parent_iterator = this;
        let current_record_promise = new Promise(function(resolve, reject) {
            parent_iterator.resolve_current_record = resolve;
            parent_iterator.reject_current_record = reject;
        });
        this.try_resolve_next_record();
        return current_record_promise;
    };


    process_record_line(line) {
        this.NR += 1;
        this.produced_records_queue.enqueue([JSON.parse(line)]);
        this.try_resolve_next_record();
    };


    process_line(line) {
        this.NL += 1;
        this.process_record_line(line);
    };


    process_data_stream_chunk(data_chunk) {
        let decoded_string = null;
        if (this.decoder) {
            try {
                decoded_string = this.decoder.decode(data_chunk);
            } catch (e) {
                if (e instanceof TypeError) {
                    this.store_or_propagate_exception(new RbqlIOHandlingError(utf_decoding_error));
                } else {
                    this.store_or_propagate_exception(e);
                }
                return;
            }
        } else {
            decoded_string = data_chunk.toString(this.encoding);
        }
        let line_starts_with_lf = decoded_string.length && decoded_string[0] == '\n';
        let first_line_index = line_starts_with_lf && this.partially_decoded_line_ends_with_cr ? 1 : 0;
        this.partially_decoded_line_ends_with_cr = decoded_string.length && decoded_string[decoded_string.length - 1] == '\r';
        let lines = csv_utils.split_lines(decoded_string);
        lines[0] = this.partially_decoded_line + lines[0];
        assert(first_line_index == 0 || lines[0].length == 0);
        this.partially_decoded_line = lines.pop();
        for (let i = first_line_index; i < lines.length; i++) {
            this.process_line(lines[i]);
        }
    };


    process_data_stream_end() {
        this.input_exhausted = true;
        if (this.partially_decoded_line.length) {
            let last_line = this.partially_decoded_line;
            this.partially_decoded_line = '';
            this.process_line(last_line);
        }
        this.try_resolve_next_record();
    };


    stop() {
        if (this.stream)
            this.stream.destroy(); // TODO consider using pause() instead
    };


    async start() {
        if (this.started)
            return;
        this.started = true;
        this.stream.on('data', (data_chunk) => { this.process_data_stream_chunk(data_chunk); });
        this.stream.on('end', () => { this.process_data_stream_end(); });
    };


    get_warnings() {
        let result = [];
        return result;
    };
}

async function query_json(query_text, input_path, output_path, output_warnings, user_init_code='') {
    let input_stream = input_path === null ? process.stdin : fs.createReadStream(input_path);
    let [output_stream, close_output_on_finish] = output_path === null ? [process.stdout, false] : [fs.createWriteStream(output_path), true];

    let default_init_source_path = path.join(os.homedir(), '.rbql_init_source.js');
    if (user_init_code == '' && fs.existsSync(default_init_source_path)) {
        user_init_code = read_user_init_code(default_init_source_path);
    }
    let input_file_dir = input_path ? path.dirname(input_path) : null;
    let join_tables_registry = null;
    let input_iterator = new JsonLinesRecordIterator(input_stream);
    let output_writer = new JsonLinesWriter(output_stream, close_output_on_finish);
    await rbql.query(query_text, input_iterator, output_writer, output_warnings, join_tables_registry, user_init_code);
}


module.exports.JsonLinesWriter = JsonLinesWriter;
module.exports.JsonLinesRecordIterator = JsonLinesRecordIterator;
module.exports.query_json = query_json;
